sdks/python/apache_beam/io/hdfilesystem.py

```diff
@@ -0,0 +1,190 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""FileSystem implementation for accessing Hadoop Distributed File System
+files."""
+
+from __future__ import absolute_import
+
+import posixpath
+
+from hdfs3 import HDFileSystem as HDFS3
+
+from apache_beam.io.filesystem import BeamIOError
+from apache_beam.io.filesystem import CompressedFile
+from apache_beam.io.filesystem import CompressionTypes
+from apache_beam.io.filesystem import FileMetadata
+from apache_beam.io.filesystem import FileSystem
+from apache_beam.io.filesystem import MatchResult
+
+__all__ = ['HdFileSystem']
+
+
+COPY_BUFFER_SIZE = 2**16
+
+
+# TODO(udim): Add @retry.with_exponential_backoff to some functions, like in
+# gcsio.py.
+
+
+class HdFileSystem(FileSystem):
+  """FileSystem implementation that supports HDFS."""
+
+  def __init__(self, *args, **kwargs):
+    """Initialize a connection to HDFS.
+
+    Keyword arguments:
+      host: Hostname of HDFS server to connect to.
+      port: Port of HDFS server to connect to.
+    """
+    host = kwargs.pop('host')
+    port = kwargs.pop('port')
+    super(HdFileSystem, self).__init__(*args, **kwargs)
+    self._hdfs = HDFS3(host=host, port=port)
+
+  @classmethod
+  def scheme(cls):
+    return 'hdfs'
+
+  def join(self, basepath, *paths):
+    return posixpath.join(basepath, *paths)
+
+  def split(self, path):
+    return posixpath.split(path)
+
+  def mkdirs(self, path):
+    if self.exists(path):
+      raise IOError('%s: path already exists' % path)
+    self._hdfs.makedirs(path)
+
+  def match(self, patterns, limits=None):
+    if limits is None:
+      limits = [None] * len(patterns)
+    else:
+      err_msg = 'Patterns and limits should be equal in length'
+      assert len(patterns) == len(limits), err_msg
+
+    def _match(pattern, limit):
+      """Find all matching paths to the pattern provided."""
+      fileinfos = self._hdfs.ls(pattern, detail=True)[:limit]
```

Comment:
<!--thread_id:cc_156249074_t; commit:3c708ac10af4457fe46e07503f0b72930832db07; resolved:1-->
<!--section:context-quote-->
> **chamikaramj** wrote:
> Does client support a method to do this more efficiently (without receiving the full match result in the form of a single list ? For example, gcsio client returns the response in the form of a batch and a token and keeps requesting batches until limit is reached. If client does not have such a method please add a TODO to update the client to perform match() invocation more efficiently.

<!--section:body-->
It seems like the underlying C++ library only supports listing the whole directory using hdfsListDirectory:
https://github.com/ContinuumIO/libhdfs3-downstream/blob/master/libhdfs3/src/client/hdfs.h#L558